File: ch3/clip2/boj1916/Main.java
import java.util.Scanner;

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        final int INF = 1000000000;
        int n = sc.nextInt(), e = sc.nextInt();
        int[][] graph = new int[n + 1][n + 1];
        int[] cost = new int[n + 1];
        boolean[] visited = new boolean[n + 1];

        for(int i = 0; i < n + 1; i++) {
            cost[i] = INF;
            for(int j = 0; j < n + 1; j++) {
                graph[i][j] = INF;
            }
        }
        for (int i = 0; i < e; i++) {
            int s = sc.nextInt(), d = sc.nextInt(), c = sc.nextInt();
            if(graph[s][d] > c) graph[s][d] = c;
        }
        int start = sc.nextInt(), end = sc.nextInt();
        cost[start] = 0;
        for(int i = 1; i <= n; i++) {
            int min = INF;
            int minIndex = -1;
            for(int j = 1; j <= n; j++) {
                if(cost[j] < min && !visited[j]) {
                    min = cost[j];
                    minIndex = j;
                }
            }
            if(minIndex == -1) break;

            visited[minIndex] = true;
            for(int j = 1; j <= n; j++) {
                if(cost[j] > cost[minIndex] + graph[minIndex][j]) {
                    cost[j] = cost[minIndex] + graph[minIndex][j];
                }
            }
        }
        System.out.println(cost[end]);
    }
}

File: ch3/clip2/boj1916/main.py
import sys

INF = 1000000000

n = int(sys.stdin.readline())
e = int(sys.stdin.readline())

graph = [[INF] * (n + 1) for _ in range(n + 1)]
cost = [INF] * (n + 1)
visited = [False] * (n + 1)

for i in range(n + 1):
    cost[i] = INF

for i in range(e):
    s, d, c = map(int, sys.stdin.readline().split())
    if graph[s][d] > c:
        graph[s][d] = c

start, end = map(int, sys.stdin.readline().split())
cost[start] = 0

for i in range(1, n + 1):
    min_cost = INF
    min_index = -1
    for j in range(1, n + 1):
        if cost[j] < min_cost and not visited[j]:
            min_cost = cost[j]
            min_index = j
    if min_index == -1:
        break

    visited[min_index] = True
    for j in range(1, n + 1):
        if cost[j] > cost[min_index] + graph[min_index][j]:
            cost[j] = cost[min_index] + graph[min_index][j]

print(cost[end])
